hphp/runtime/vm/jit/vasm-reuseimm.cpp

```diff
@@ -0,0 +1,182 @@
+/*
+   +----------------------------------------------------------------------+
+   | HipHop for PHP                                                       |
+   +----------------------------------------------------------------------+
+   | Copyright (c) 2010-present Facebook, Inc. (http://www.facebook.com)  |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 3.01 of the PHP license,      |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | http://www.php.net/license/3_01.txt                                  |
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+*/
+
+#include <cstdlib>
+
+#include "hphp/runtime/vm/jit/vasm.h"
+
+#include "hphp/runtime/vm/jit/vasm-gen.h"
+#include "hphp/runtime/vm/jit/vasm-instr.h"
+#include "hphp/runtime/vm/jit/vasm-print.h"
+#include "hphp/runtime/vm/jit/vasm-reg.h"
+#include "hphp/runtime/vm/jit/vasm-unit.h"
+#include "hphp/runtime/vm/jit/vasm-util.h"
+
+#include "hphp/util/assertions.h"
+
+TRACE_SET_MOD(vasm);
+
+namespace HPHP { namespace jit {
+
+struct qTrack {
+  Immed64 val;
+  Vreg base;
+  int  ttl;
+};
+
+struct Env {
+  Vunit& unit;
+  int tick;
+  std::deque<qTrack> qdeque; // leaky bucket of recent immediates
+};
+
+static bool isMultiword(int64_t imm) {
+  uint64_t val = abs(imm);
+  if (val > (1 << 16)) return true;
+  return false;
+}
+
+// candidate around +/- uimm12
+static bool reuseCandidate(Env& env, int64_t p, Vreg& reg, int& delta)
+{
+  delta = 0;
+  for (const auto& itr : env.qdeque) {
+    int64_t q = itr.val.q();
+    if ((p >= q) && (p < (q+4095))) {
+      delta = safe_cast<int>(p-q);
+      reg = itr.base;
+      return true;
+    }
+    if ((p < q) && (q < (p+4095))) {
+      delta = safe_cast<int>(q-p);
+      delta = -delta;
+      reg = itr.base;
+      return true;
+    }
+  }
+  return false;
+}
+
+template <typename Inst>
+void reuseImmq(Env& env, const Inst& inst, Vlabel b, size_t i) { return; }
+
+// purge table across any instruction with RegSet or flow change
```

Comment:
"flow change"  control flow change.

I'm not sure about your whitelist comment.  This is the same technique used in
other places in the compiler, just using the X-macro pattern.  Can you elaborate?